src/node/hooks/express/openapi.js

```diff
@@ -0,0 +1,666 @@
+            timestamp: {
+              type: 'integer',
+            },
+          },
+        },
+        Message: {
+          type: 'object',
+          properties: {
+            text: {
+              type: 'string',
+            },
+            userId: {
+              type: 'string',
+            },
+            userName: {
+              type: 'string',
+            },
+            time: {
+              type: 'integer',
+            },
+          },
+        },
+      },
+      responses: {
+        ...defaultResponses,
+      },
+      securitySchemes: {
+        ApiKey: {
+          type: 'apiKey',
+          in: 'query',
+          name: 'apikey',
+        },
+      },
+    },
+    security: [{ ApiKey: [] }],
+  };
+
+  // build operations
+  for (const funcName in apiHandler.version[version]) {
+    let operation = {};
+    if (operations[funcName]) {
+      operation = { ...operations[funcName] };
+    } else {
+      // console.warn(`No operation found for function: ${funcName}`);
+      operation = {
+        operationId: funcName,
+        responses: defaultResponseRefs,
+      };
+    }
+
+    // set parameters
+    operation.parameters = operation.parameters || [];
+    for (const paramName of apiHandler.version[version][funcName]) {
+      operation.parameters.push({ $ref: `#/components/parameters/${paramName}` });
+      if (!definition.components.parameters[paramName]) {
+        definition.components.parameters[paramName] = {
+          name: paramName,
+          in: 'query',
+          schema: {
+            type: 'string',
+          },
+        };
+      }
+    }
+
+    // set path
+    let path = `/${operation.operationId}`; // APIPathStyle.FLAT
+    if (style === APIPathStyle.REST && operation._restPath) {
+      path = operation._restPath;
+    }
+    delete operation._restPath;
+
+    // add to definition
+    // NOTE: It may be confusing that every operation can be called with both GET and POST
+    definition.paths[path] = {
+      get: {
+        ...operation,
+        operationId: `${operation.operationId}UsingGET`,
+      },
+      post: {
+        ...operation,
+        operationId: `${operation.operationId}UsingPOST`,
+      },
+    };
+  }
+  return definition;
+};
+
+exports.expressCreateServer = async (_, args) => {
+  const { app } = args;
+
+  // create openapi-backend handlers for each api version under /api/{version}/*
+  for (const version in apiHandler.version) {
+    // we support two different styles of api: flat + rest
+    // TODO: do we really want to support both?
+
+    for (const style of [APIPathStyle.FLAT, APIPathStyle.REST]) {
+      const apiRoot = getApiRootForVersion(version, style);
+
+      // generate openapi definition for this API version
+      const definition = generateDefinitionForVersion(version, style);
+
+      // serve version specific openapi definition
+      app.get(`${apiRoot}/openapi.json`, (req, res) => {
+        res.header('Access-Control-Allow-Origin', '*');
```

Comment:
Potential big red flag here.

I am not really sure allowing CORS from everywhere would pass a security assessment.

Probably this needs to be configurable in the settings (and, jost out of my head, we have to test embedding works afterwards).

More work needed here.

Reply:
For exposing the OAS definitions allowing CORS from everywhere is probably ok.